src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -200,10 +295,103 @@ Score(TreeType& queryNode, TreeType& referenceNode)
     }
     score = DBL_MAX;
+
+    // Store not used alpha for Monte Carlo.
+    if (kernelIsGaussian && monteCarlo)
+      queryStat.AccumAlpha() += CalculateAlpha(referenceNode.Stat().Depth());
+  }
+  else if (monteCarlo &&
+           referenceNode.NumDescendants() >= mcAccessCoef * initialSampleSize &&
+           kernelIsGaussian)
+  {
+    // Calculate reference node depth.
+    if (referenceStat.Depth() == 0)
+      referenceStat.Depth() = CalculateDepth(referenceNode);
+
+    // Monte Carlo probabilistic estimation.
+    // Calculate alpha.
+    const double depthAlpha = CalculateAlpha(referenceNode.Stat().Depth());
+    double alpha = depthAlpha + queryStat.AccumAlpha();
+    if (alpha > 1)
+      alpha = 1;
+
+    const boost::math::normal normalDist;
+    const double z =
+        std::abs(boost::math::quantile(normalDist, alpha / 2));
+    const size_t numDesc = referenceNode.NumDescendants();
+    arma::vec sample;
+    arma::vec means = arma::zeros(queryNode.NumDescendants());
+    size_t m;
+    double meanSample;
+    bool useMonteCarloPredictions = true;
+    for (size_t i = 0; i < queryNode.NumDescendants(); ++i)
+    {
+      const size_t queryIndex = queryNode.Descendant(i);
+      sample.clear();
+      m = initialSampleSize;
+      while (m > 0)
+      {
+        const size_t oldSize = sample.size();
+        const size_t newSize = oldSize + m;
+
+        // Don't use probabilistic estimation if this is gonna take a close
+        // amount of computation to the exact calculation.
+        if (newSize >= mcBreakCoef * numDesc)
+        {
+          useMonteCarloPredictions = false;
+          break;
+        }
+
+        // Increase the sample size.
+        sample.resize(newSize);
+        for (size_t i = 0; i < m; ++i)
+        {
+          // Sample and evaluate random points from the reference node.
+          const size_t randomPoint = math::RandInt(0, numDesc);
+          sample(oldSize + i) =
+              EvaluateKernel(queryIndex, referenceNode.Descendant(randomPoint));
+        }
+        meanSample = arma::mean(sample);
+        const double stddev = arma::stddev(sample);
+        const double mThreshBase =
+            z * stddev * (1 + relError) / (relError * meanSample);
+        const size_t mThresh = std::ceil(mThreshBase * mThreshBase);
+
+        if (sample.size() < mThresh)
+          m = mThresh - sample.size();
+        else
+          m = 0;
+      }
+
+      if (useMonteCarloPredictions)
+        means(i) = meanSample;
+      else
+        break;
+    }
+
+    if (useMonteCarloPredictions)
+    {
+      // Use Monte Carlo estimation.
+      queryStat.AccumAlpha() = 0;
+      score = DBL_MAX;
+      for (size_t i = 0; i < queryNode.NumDescendants(); ++i)
+        densities(queryNode.Descendant(i)) += numDesc * means(i);
+    }
+    else
+    {
+      // Reclaim not used alpha for Monte Carlo.
+      queryStat.AccumAlpha() += CalculateAlpha(referenceNode.Stat().Depth());
```

Comment:
I think the bug may have to do with this---here you are adding to `AccumAlpha()` at every non-approximated reference node.  Let's say I add the alpha for a node to `AccumAlpha()` here, but then I recurse (say there are two children) and those children, too, do not use Monte Carlo approximation.  Then, we will add the alpha for each of the children, but we already know that the sum of the two child alphas is the parent alpha---so, overall, we've added 2*alpha, which is twice as much as we should have.  I think this may be the reason for the too-large alpha values.

Reply:
Solved with f0aa09d651007b0c12f7b01f724f959f776b3e52 and e4d507d4507218d5c2051aa538e9014c587c9dd8.